### PR TITLE
test publisher/subscription with the c/cpp typesupport for test_msgs::msg::array

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
From the doc below, it seems better to add tests for them.

https://github.com/ros2/rcl/blob/c5e07c1d96bd44b304fd1c9f92a87ff914e2c205/rcl/include/rcl/publisher.h#L82-L96
and
https://github.com/ros2/rcl/blob/67919d56192f12c96147bcc1d459f9aeea880b3f/rcl/include/rcl/subscription.h#L89-L104

### CODE REVIEW

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18838/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13368/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19588/)

**Comment by clalancette** [comment]
The failed test on aarch64 is a known flake, so no worries there.

The failures on Windows are not known, but I also have a hard time seeing how this PR is involved.  I'm going to re-run it:


* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19633/)

**Comment by iuhilnehc-ynos** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18900/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13435/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19657/)

File: rcl/test/rcl/test_subscription.cpp [hunk 1]
```diff
@@ -1741,3 +1757,124 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_init_fini_ma
...
+  {
+    param_ = GetParam();
+    TEST_FIXTURE_P_RMW(TestSubscriptionFixture) ::SetUp();
+  }
+
+  TestParameters param_;
+};
+
+
+/* Test subscription to recevie complex message from a publisher with typesupport settings.
```
Comment (clalancette) [inline rcl/test/rcl/test_subscription.cpp]:
```suggestion
/* Test subscription to receive complex message from a publisher with typesupport settings.
```

File: rcl/test/rcl/test_subscription.cpp [hunk 2]
```diff
@@ -1741,3 +1757,124 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_init_fini_ma
...
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_subscription_fini(&subscription, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+  ASSERT_TRUE(wait_for_established_subscription(&publisher, 10, 100));
+  constexpr char test_string[] = "testing";
+  constexpr bool bool_values[3] = {true, false, true};
+  {
```
Comment (clalancette) [inline rcl/test/rcl/test_subscription.cpp]: Any particular reason for this extra scope?

File: rcl/test/rcl/test_subscription.cpp [hunk 3]
```diff
@@ -1741,3 +1757,124 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_init_fini_ma
...
+    } else {
+      test_msgs::msg::Arrays msg;
+      std::copy(std::begin(bool_values), std::end(bool_values), msg.bool_values.begin());
+      msg.string_values[1] = test_string;
+      ret = rcl_publish(&publisher, &msg, nullptr);
+    }
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription, context_ptr, 10, 100));
+  {
```
Comment (clalancette) [inline rcl/test/rcl/test_subscription.cpp]: Any particular reason for this extra scope?

**Review by clalancette — COMMENTED** [review]
I've left a couple of minor comments.

File: rcl/test/rcl/test_subscription.cpp (hunk 2 above)
Comment (iuhilnehc-ynos) [reply rcl/test/rcl/test_subscription.cpp]: To be honest, there's no particular reason. I just wanted to make the `rcl_publish / rcl_take` action inside a more clear block. 

I removed it in https://github.com/ros2/rcl/pull/1074/commits/d37e838d8a2f07c049838a36e891d39e854aa464.

File: rcl/test/rcl/test_subscription.cpp (hunk 3 above)
Comment (iuhilnehc-ynos) [reply rcl/test/rcl/test_subscription.cpp]: removed it in https://github.com/ros2/rcl/commit/d37e838d8a2f07c049838a36e891d39e854aa464

File: rcl/test/rcl/test_subscription.cpp (hunk 1 above)
Comment (iuhilnehc-ynos) [reply rcl/test/rcl/test_subscription.cpp]: Thanks.

File: rcl/test/CMakeLists.txt [hunk 4]
```diff
@@ -6,6 +6,7 @@ find_package(osrf_testing_tools_cpp REQUIRED)
 find_package(rcpputils REQUIRED)
 find_package(rcutils REQUIRED)
 find_package(rmw_implementation_cmake REQUIRED)
+find_package(rosidl_runtime_cpp REQUIRED)
```
Comment (clalancette) [inline rcl/test/CMakeLists.txt:9]: For future readers; I was slightly concerned that this was going to introduce a new (test) dependency on this package, but it turns out that `rcl` already indirectly depends on `rosidl_runtime_cpp`, so no worries there.

**Review by clalancette — APPROVED** [review]
Looks good with green CI.

**Comment by iuhilnehc-ynos** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18923/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13461/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19676/)

**Pull request merged**